packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -86,10 +82,18 @@ export default function astro({ settings, logger }: AstroPluginOptions): vite.Pl
 
 			// Astro scripts and styles virtual module code comes from the main Astro compilation
 			// through the metadata from `astroFileToCompileMetadata`. It should always exist as Astro
-			// modules are compiled first, then its virtual modules. If the virtual modules are somehow
-			// compiled first, throw an error and we should investigate it.
+			// modules are compiled first, then its virtual modules.
 			const filename = normalizePath(normalizeFilename(parsedId.filename, config.root));
-			const compileMetadata = astroFileToCompileMetadata.get(filename);
+			let compileMetadata = astroFileToCompileMetadata.get(filename);
+			// If `compileMetadata` doesn't exist in dev, that means the virtual module may have been invalidated.
+			// We try to re-compile the main Astro module (`filename`) first before retrieving the metadata again.
+			if (!compileMetadata && server) {
+				const code = await loadId(server.pluginContainer, filename);
+				if (code != null) await compile(code, filename);
+				compileMetadata = astroFileToCompileMetadata.get(filename);
```

Comment:
Who does add data to `astroFileToCompileMetadata`? The function `loadId` or `compile`?

Reply:
`compile` will populate `astroFileToCompileMetadata`.